pkg/signature/kms/cliplugin/README.md

```diff
@@ -0,0 +1,91 @@
+# CLI Plugin
+
+This is a package and module for using custom KMS plugins as separate executables.
+It is intended to be used by cosign, but you may use this in your own programs that import sigstore.
+
+## Design
+
+We follow [kubectl's style](https://kubernetes.io/docs/tasks/extend-kubectl/kubectl-plugins/#writing-kubectl-plugins) of plugins. Any language that can create CLI programs can be used to make a plugin.
+
+### Usage
+
+Plugins are separate programs on your system's PATH, named in the scheme `sigstore-kms-[name]`, like `sigstore-kms-my-hsm`. They can be invoked with cosign like `cosign [sub-command] --key "my-hsm://my-key-id" ...
+`
+
+### Protocol
+
+The main program will invoke the program with these specifications:
+
+* stdin
+  * Data to be signed or verified.
+* arg 1
+  * A number identifying the version of this protocol.
+    * In the future we may change this protocol, either the encoding or the formatting of argument and return values.
+    Protocol changes will result in a major-version bump for the library.
+    When changing the protocol, new versions of sigstore will not maintain backwards compatibility with
+    previous protocol versions. If a plugin author wishes, they may branch their plugin program’s behaviour
+    to be compatible with multiple versions of the protocol, or multiple major versions of sigstore and cosign.
+* arg 2
+  * JSON of initialization options and method arguments.
+* stdout
+  * JSON of method return values.
+
+See [./common/interface.go](./common/interface.go) for the full JSON schema.
+
+The plugin program must first exit before sigstore begins parsing responses.
```

Comment:
why is that? In general I dont think you could enforce that

Reply:
Currently, we invoke the plugin program with `Cmd.Output()`, which waits until the plugin program exits, because:
1. We need to make sure that the main program received all of the intended output from the plugin program.
2. For the sake of users we may want to guarantee that plugin programs do eventually exit.

If we don't wait for the plugin program to exit, the protocol will need to be slightly more complex: read a full line of stdout before parsing the `PluginResp`. The newline character serves as a delimeter.